Tools/cases_generator/stacking.py

```diff
@@ -103,16 +102,23 @@ class StackItem:
     offset: StackOffset
     effect: StackEffect
 
-    def as_variable(self) -> str:
+    def as_variable(self, lax: bool = False) -> str:
         """Return e.g. stack_pointer[-1]."""
         terms = self.offset.as_terms()
         if self.effect.size:
             terms.insert(0, ("+", "stack_pointer"))
         index = make_index(terms)
         if self.effect.size:
-            return index
+            res = index
         else:
-            return f"stack_pointer[{index}]"
+            res = f"stack_pointer[{index}]"
+        if not lax:
+            # Check that we're not reading or writing above stack top.
+            # Skip this for output variable initialization (lax=True).
```

Comment:
I'm still pondering if I can add a working check for writing arrays above stack level. This is tricky because output array variables are initialized before the body of the opcode, at a point where `stack_pointer` is still low. E.g. 
```
        inst(UNPACK_SEQUENCE_TUPLE, (unused/1, seq -- values[oparg])) {
            ...
        }
```
which produces output like this:
```
        TARGET(UNPACK_SEQUENCE_TUPLE) {
            PyObject *seq;
            PyObject **values;
            seq = stack_pointer[-1];
            values = stack_pointer - 1;
            ...
            STACK_SHRINK(1);
            STACK_GROW(oparg);
            next_instr += 1;
            DISPATCH();
        }
```
The assert would trigger, because the code (`...`) may well write above `stack_pointer`, but all is well because of the following `STACK_GROW(oparg)`. Hence the `lax` flag, which disables the assert for output array variables (L372 below). It's hard to conceive of a realistic example where the failing assert would not be a false positive. A theoretical example would be:
```
op(A, (-- temp[oparg])) { ... }
op(B, (temp[oparg] --) { ... }
macro(M) = A + B;
```
But I don't expect we'll ever write such code.